lib/reek/context/visibility_tracker.rb

```diff
@@ -0,0 +1,54 @@
+require 'private_attr/everywhere'
+
+module Reek
+  module Context
+    # Responsible for tracking visibilities in regards to CodeContexts.
+    # :reek:Attribute
+    class VisibilityTracker
+      attr_accessor :visibility
+      private_attr_accessor :tracked_visibility
+
+      def initialize(visibility = :public)
+        @visibility = visibility
+      end
+
+      # Handle the effects of a visibility modifier.
+      #
+      # @example Modifying the visibility of existing children
+      #   track_visibility children, :private, [:hide_me, :implementation_detail]
+      #
+      # @param children [Array<CodeContext>]
+      # @param visibility [Symbol]
+      # @param names [Array<Symbol>]
+      #
+      def track_visibility(children, visibility, names)
+        if names.any?
+          children.each do |child|
+            child.visibility = visibility if names.include?(child.name)
+          end
+        else
+          self.tracked_visibility = visibility
+        end
+      end
```

Comment:
>> Hm, this method is called exactly once (in CodeContext#track_visibility), and the caller knows whether names is empty or not (and could just as well filter children based on names on the caller side), so this smells to me of control parameter.
This also means that the second half of this method (when names is #empty?) ignores ⅔ of the method’s arguments and uses only the middle one. 

I agree that this method is not the best one! Let's improve on that in subsequent pull requests, I'm rampaging over CodeContexts and friends right now heavily anyways!

>>  I’d at least make them keyword arguments (sadly, we can’t make them required without the foo: raise hack while we support Ruby 2.0).

Updated.